### PR TITLE
CLOUDP-237545: Ignore tenant readonly fields for Kubernetes generation

### PR DESCRIPTION
**Pull request by roothorp**
## Proposed changes

When generating configuration/YAML for the Atlas Kubernetes Operator via the CLI, and the deployment is a shared/tenant cluster, read only fields (such as biConnector) are still exported and the operator tries to apply them. This results in an error from the Atlas API, as readonly attributes cannot be set.

This change means that when we are generating configuration for a shared cluster, we remove the readonly fields to avoid this issue; replicating the logic already used in the CLI [here](https://github.com/mongodb/mongodb-atlas-cli/blob/master/internal/cli/atlas/clusters/clusters.go#L82).

_Jira ticket:_ [CLOUDP-237545](https://jira.mongodb.org/browse/CLOUDP-237545)

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [ ] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/kubernetes/operator/deployment/deployment.go [hunk 1]
```diff
@@ -171,9 +171,36 @@ func BuildAtlasAdvancedDeployment(deploymentStore store.OperatorClusterStore, va
...
+
+	if isTenant {
+		out.DeploymentSpec.BiConnector = nil
+		out.DeploymentSpec.EncryptionAtRestProvider = ""
+		out.DeploymentSpec.DiskSizeGB = nil
+		out.DeploymentSpec.MongoDBMajorVersion = ""
+		out.DeploymentSpec.PitEnabled = nil
+		out.DeploymentSpec.BackupEnabled = nil
+	}
+}
```
Comment (s-urbaniak) [inline internal/kubernetes/operator/deployment/deployment.go:202]: This misses a couple of nil checks and can panic. Suggestion:

```
...
	if hasTenantFields(&atlasDeployment.Spec) {
		atlasDeployment.Spec.DeploymentSpec.BiConnector = nil
		atlasDeployment.Spec.DeploymentSpec.EncryptionAtRestProvider = ""
		atlasDeployment.Spec.DeploymentSpec.DiskSizeGB = nil
		atlasDeployment.Spec.DeploymentSpec.MongoDBMajorVersion = ""
		atlasDeployment.Spec.DeploymentSpec.PitEnabled = nil
		atlasDeployment.Spec.DeploymentSpec.BackupEnabled = nil
	}

	return deploymentResult, nil
}

func hasTenantFields(out *akov2.AtlasDeploymentSpec) bool {
	if out.DeploymentSpec == nil {
		return false
	}
	for _, spec := range out.DeploymentSpec.ReplicationSpecs {
		if spec == nil {
			continue
		}
		for _, c := range spec.RegionConfigs {
			if c == nil {
				continue
			}
			if c.ProviderName == "TENANT" {
				return true
			}
		}
	}
	return false
}
```

**Review by s-urbaniak — APPROVED** [review]
see comment

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
one change please as this can panic.

File: internal/kubernetes/operator/deployment/deployment_test.go [hunk 2]
```diff
@@ -520,3 +520,51 @@ func TestBuildServerlessDeployments(t *testing.T) {
...
+				{
+					RegionConfigs: []*akov2.AdvancedRegionConfig{
+						{
+							ProviderName: "TENANT",
+						},
+					},
+				},
+			},
+		},
+	}
```
Comment (s-urbaniak) [inline internal/kubernetes/operator/deployment/deployment_test.go:545]: make this a table test with a couple of cases:
1. DeploymentSpec is nil
2. ReplicationSpecs is nil
3. RegionConfigs is nil
4. RegionConfigs has multiple entries which does not include any "TENANT"
5. RegionConfigs has multiple entries which does include "TENANT"

File: internal/kubernetes/operator/deployment/deployment.go (hunk 1 above)
Comment (roothorp) [reply internal/kubernetes/operator/deployment/deployment.go:202]: How can `spec` or `c` be nil if we're ranging over them? If there's nothing there, we won't enter the loop, right?

Comment (s-urbaniak) [reply internal/kubernetes/operator/deployment/deployment.go:202]: @roothorp the following two cases panic with this code:
```
	spec := akov2.AtlasDeploymentSpec{
		DeploymentSpec: &akov2.AdvancedDeploymentSpec{
			ReplicationSpecs: []*akov2.AdvancedReplicationSpec{
				{
					nil,
				},
			},
		},
	}

	cleanTenantFields(&spec)
```

Comment (s-urbaniak) [reply internal/kubernetes/operator/deployment/deployment.go:202]:
```
	spec := akov2.AtlasDeploymentSpec{
		DeploymentSpec: &akov2.AdvancedDeploymentSpec{
			ReplicationSpecs: []*akov2.AdvancedReplicationSpec{
				{
					RegionConfigs: []*akov2.AdvancedRegionConfig{
						nil,
					},
				},
			},
		},
	}

	cleanTenantFields(&spec)
```

File: internal/kubernetes/operator/deployment/deployment_test.go (hunk 2 above)
Comment (roothorp) [reply internal/kubernetes/operator/deployment/deployment_test.go]: done!

File: internal/kubernetes/operator/deployment/deployment.go (hunk 1 above)
Comment (roothorp) [reply internal/kubernetes/operator/deployment/deployment.go:201]: done!

File: internal/kubernetes/operator/deployment/deployment_test.go [hunk 3]
```diff
@@ -520,3 +520,105 @@ func TestBuildServerlessDeployments(t *testing.T) {
...
+				},
+			},
+			expect: true,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := hasTenantRegionConfig(&akov2.AtlasDeployment{
+				Spec: tt.spec,
+			}); got != tt.expect {
+				t.Errorf("")
```
Comment (s-urbaniak) [inline internal/kubernetes/operator/deployment/deployment_test.go:620]:
```suggestion
				t.Errorf("expect hasTenantRegionConfig to be %t, got %t", tt.expect, got)
```

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
one change, else LGTM

**Review by s-urbaniak — APPROVED** [review]

**Review by gssbzn — APPROVED** [review]

**Pull request merged**